### PR TITLE
Added support for NTAG 21x tags

### PR DESCRIPTION
**Pull request by SloCompTech**
Hello,
I added support for __NTAG21x__ tags according to NTAG213/215/216 documentation Rev. 3.2 June 2015
It includes:
- reading pages
- writting to pages
- authentication with tag
- configuring counters

Before it can be merged some things need to be checked:
1. I added myself in copyright in files I writen or I updated, so authors need to give __permission__ that they agree with changes
2. Please check if I correctly updated compiler settings for ntag21x.c and examples

### CODE REVIEW

File: libfreefare/freefare.c [hunk 1]
```diff
@@ -37,19 +37,21 @@ freefare_tag_new (nfc_device *device, nfc_target target)
...
-	tag = mifare_desfire_tag_new (device, target);
+	     tag = mifare_desfire_tag_new (device, target);
+    } else if (ntag21x_taste (device, target)) {
+ 	     tag = ntag21x_tag_new (device, target);
     } else if (mifare_ultralightc_taste (device, target)) {
-	tag = mifare_ultralightc_tag_new (device, target);
+	     tag = mifare_ultralightc_tag_new (device, target);
     } else if (mifare_ultralight_taste (device, target)) {
-	tag = mifare_ultralight_tag_new (device, target);
+	     tag = mifare_ultralight_tag_new (device, target);
```
Comment (smortex) [inline libfreefare/freefare.c]: Indentation changed unexpectedly

File: libfreefare/freefare.c [hunk 2]
```diff
@@ -169,6 +171,8 @@ freefare_get_tag_friendly_name (FreefareTag tag)
 	return "Mifare UltraLightC";
     case MIFARE_ULTRALIGHT:
 	return "Mifare UltraLight";
+    case NTAG_21x:
+  return "NTAG21x";
```
Comment (smortex) [inline libfreefare/freefare.c]: Indentation look inconsistent

File: libfreefare/freefare.h [hunk 3]
```diff
@@ -102,7 +106,54 @@ bool		 is_mifare_ultralight (FreefareTag tag);
...
+int ntag21x_read4(FreefareTag tag,uint8_t page,uint8_t *data); // Read 4 bytes on page
+int ntag21x_fast_read (FreefareTag tag, uint8_t start_page,uint8_t end_page, uint8_t *data); // Read n*4 bytes from range [start_page,end_page]
+int ntag21x_fast_read4(FreefareTag tag,uint8_t page,uint8_t *data); // Fast read certain page
+int ntag21x_read_cnt (FreefareTag tag, uint8_t *data); // Read 3-byte NFC counter if enabled else it returns error
+int ntag21x_write (FreefareTag tag, uint8_t page, uint8_t data[4]); // Write 4 bytes to page
+int ntag21x_compatibility_write (FreefareTag tag, uint8_t page, uint8_t data[4]); // Writes 4 bytes to page with mifare classic write
+int ntag21x_authenticate (FreefareTag tag, const NTAG21xKey key); // Authenticate with tag
+bool is_ntag21x(FreefareTag tag); // Check if tag type is NTAG21x
+bool ntag21x_is_auth_supported (nfc_device *device, nfc_iso14443a_info nai); // Check if tag supports 21x commands
+
```
Comment (smortex) [inline libfreefare/freefare.h]: See *Function prototypes* in the [style guide](https://www.freebsd.org/cgi/man.cgi?query=style&apropos=0&sektion=9&manpath=FreeBSD+11.0-RELEASE+and+Ports&arch=default&format=html) for the *magic* of indentation here :-)

File: libfreefare/freefare_internal.h [hunk 4]
```diff
@@ -260,7 +282,7 @@ struct mifare_ultralight_tag {
 #define MIFARE_CLASSIC(tag) ((struct mifare_classic_tag *) tag)
 #define MIFARE_DESFIRE(tag) ((struct mifare_desfire_tag *) tag)
 #define MIFARE_ULTRALIGHT(tag) ((struct mifare_ultralight_tag *) tag)
-
```
Comment (smortex) [inline libfreefare/freefare_internal.h]: Keep the blank line between this group of macros and the comment

**Comment by smortex** [comment]
Oh, and no need to worry about the copyright, #49 is about removing them all since they are never updated.

**Comment by doegox** [comment]
@smortex we've a "style" phony target in libnfc, maybe you could setup sth similar for libfreefare? https://github.com/nfc-tools/libnfc/blob/master/Makefile.am

**Comment by SloCompTech** [comment]
Fixed some identation, is there easy way to check if all is correct ?

**Comment by smortex** [comment]
Hello @SloCompTech,

> is there easy way to check if all is correct ?

In order to check style "the easy way", I simply look at the [Files changed](https://github.com/nfc-tools/libfreefare/pull/53/files) tab of the pull request (on top of this page) which shows the wholes changes of all commits combined.

 Since these changes are related to adding a new feature, the page should only contain bunches of new lines of code (new files, etc), and a few logical changes.

Therefore, any space change is suspicious and even if the spacing is wrong according to the style, it's better to keep it as it is in order to make the review easier (e.g. most changes in [libfreefare/freefare.c](https://github.com/nfc-tools/libfreefare/pull/53/files#diff-9244025360d782eb413e2c2c208eb6a3) should not be part of this pull request in order to make it easy to spot the actual changes (L74-75)).

On this tab, the in-line review comments are automatically hidden when they are attached to code that has been changed, so the comments I made some time ago and are still relevant are shown there.

Thank you for taking care of this, much appreciated :+1:

**Comment by SloCompTech** [comment]
Hi,
I think it's better that I take original repository and copy and paste changes into original version and then commit so other stuff will be unaffected.

**Comment by smortex** [comment]
@SloCompTech yes, you can create a new branch in order to have a reference to the current state of the repository, revert your `master` branch to `origin/master` and then update your `master` branch to only include the wanted changes.

Then, when you push your changes (`-f` will be required in order to force the commit since some pushed commits will have been changed) the pull request will be updated automatically (when you create a pull-request, it is linked to the branch you created, adding / modifying / removing commits in the branch updates the pull-request).

**Comment by SloCompTech** [comment]
Hi,
I used clean current repo and insert changes only and used force push, so I think it's all done.

**Comment by smortex** [comment]
Hi

I cloned the repository, rebased on top of master, and added a commit to fix some spacing problems, unfortunately I can't push it to your repository :open_mouth:.

Either you did not "Allow edits from maintainers" (checkbox at the bottom of the right column of this page), otherwise GitHub do not want me to rebase your changes and I should revert that.  Can you please check that maintainers edits are allowed and tell me?

Thansk!

**Comment by SloCompTech** [comment]
hi @smortex, I have edit by maintainers enabled, but just in case I added you as colaborator, so please try to push to repo again

**Comment by smortex** [comment]
Nah, with your authorization I realized that I was pushing to a new non-existent branch in your repo and not in master…  I just cherry-picked my commit and cleaned-up my mess!

**Comment by smortex** [comment]
Changes squashed and merged.  Thanks!

**Pull request merged**